parsing-utilities/src/main/java/gov/nasa/jpl/aerie/json/JsonParser.java

```diff
@@ -3,31 +3,130 @@
+   *
+   * @param anchors
+   *   a mutable collection of JSON Schema identifiers for previously-visited parsers
+   * @return
+   *   a JSON Schema document describing the format of JSON documents supported by this parser.
+   */
+  JsonObject getSchema(SchemaCache anchors);
+
+  /**
+   * Attempts to parse a JSON document into the domain type {@code T}.
+   *
+   * <p> This method must successfully parse any document generated by {@link #unparse(T)},
+   * and the resulting domain value must be equal to the one that produced that document.
+   * No guarantees are made for documents not generated by the {@code unparse} method.
+   * See {@link #getSchema()} for more on the relationships among these methods. </p>
+   *
+   * @param json
+   *   the JSON document to parse into a domain value
+   * @return
+   *   a wrapped domain value if parsing succeeds, or failure information otherwise
+   */
   JsonParseResult<T> parse(JsonValue json);
+
+  /**
+   * Produces a JSON document representing the given value.
+   *
+   * <p> The document produced by this method must validate against the schema returned by {@link #getSchema()},
+   * and must be parsed successfully by {@link #parse(JsonValue)} into a value equal to the one given here. </p>
+   *
+   * @param value
+   *   a domain value to encode in JSON
+   * @return
+   *   a JSON document representing the given value
+   */
   JsonValue unparse(T value);
 
+  /**
+   * Produces a JSON Schema document describing the format of JSON documents produced by {@link #unparse(T)}.
+   *
+   * <p> All documents returned from {@link #unparse(T)} must validate against this schema.
+   * However, {@link #parse(JsonValue)} is only required to faithfully parse documents generated by {@link #unparse(T)}.
+   * The {@code parse} method may reject other documents that validate relative to the schema (for instance, if the
+   * document does not maintain invariants required of the domain type), and may accept other documents that do not
+   * validate. </p>
+   *
+   * <p> Individual parsers may make stronger guarantees about the behavior of the {@code parse} method.
+   * Absent such guarantees, clients of the {@code parse} method should treat this schema as advisory only:
+   * if a document does not validate, then it definitely isn't something that {@code unparse} would produce. </p>
+   *
+   * @return
+   *   a JSON Schema document describing the format of JSON documents supported by this parser.
+   */
   default JsonObject getSchema() {
     return Json
         .createObjectBuilder()
         .add("$schema", Json.createValue("https://json-schema.org/draft/2020-12/schema"))
-        .addAll(Json.createObjectBuilder(this.getSchema(new HashMap<>())))
+        .addAll(Json.createObjectBuilder(this.getSchema(new SchemaCache())))
         .build();
   }
 
+  /**
+   * Adapts this parser losslessly from one domain type to another.
```

Comment:
In practice we don't use this losslessly :|. I believe in hasura actions we sometimes map `string -> UNIT` and `UNIT -> ""` because its only used for deserialization and we don't care about the field, only that it exists. I guess that's not a issue with the docs though.

Reply:
Nah, hang on, I see the point you're making. Let me stew on this.

Reply:
You're right. Looking at this one for instance:

```java
  public static final JsonParser<HasuraAction<HasuraAction.ActivityInput>> hasuraActivityActionP
      = hasuraActionP(hasuraActivityInputP)
      . map(Iso.of(
          untuple((name, input, session, requestQuery) -> new HasuraAction<>(name, input, session)),
          $ -> tuple($.name(), $.input(), $.session(), "")));
````

The parser is perfectly well-formed, as far as I can tell. An instance of `HasuraAction<HasuraAction.ActivityInput>` should unparse and then parse back to an equivalent instance. This is not true of parsing then unparsing (which will replace `requestQuery` with an empty string), but as noted, it doesn't need to be true.

That means that "isomorphism" is not the correct notion here. We just want the parsing direction to undo the unparsing direction -- that is, parsing is a "left inverse" of unparsing. The notion of isomorphism specializes to a two-sided inverse, but that's too strong for us.

I'll make some changes and come up with a better name.

Reply:
I've renamed `Iso` to `Convert`, adjusted its description to reflect the weaker requirements, and (as a bonus) replaced all uses of `map(Iso.of(x, y))` with a simpler `map(x, y)` form.